### PR TITLE
[2] Add support for periodic builds

### PR DESCRIPTION
**Pull request by pranavrc**
# Context

https://github.com/screwdriver-cd/screwdriver/issues/688

This PR adds support for periodic builds in executor-queue. It includes code merged from https://github.com/screwdriver-cd/executor-queue/pull/13 (PR to transform jenkins-style cron to standard GNU cron).

The function _startPeriodic_() takes a config object from the jobFactory model, sets this config in redis as {jobid: config} and enqueues a delayed job (with a timestamp computed based on the cron) in the redis queue, defined as _startDelayed_ in the instantiated multiWorker. The _startDelayed_ in the multiWorker then gets the stored config in redis and calls back _startPeriodic_ with a flag _trigger=true_. This flag causes _startPeriodic_ to post a build event to the API with the pipeline ID and job name, causing a build to trigger in the pipeline. After posting the event, _startPeriodic_ enqueues another delayed job in the queue with a newly computed timestamp as part of its normal logic.

The function _stopPeriodic_() takes a config object with jobID, removes the entry from the redis table, and deletes existing jobs from the queue with that job name.

If an isUpdate flag is passed to _startPeriodic_, it will call _stopPeriodic_ to kill existing jobs before proceeding with its usual logic.

# Objective

- [x] Cron helper implementation to get next timestamp based on cron annotation
- [x] startPeriodic() implementation
- [x] stopPeriodic() implementation
- [x] Unit tests

# Related links

API PR - https://github.com/screwdriver-cd/screwdriver/pull/1032
Models PR - https://github.com/screwdriver-cd/models/pull/239
Queue Worker PR - https://github.com/screwdriver-cd/queue-worker/pull/57
Executor Base PR - https://github.com/screwdriver-cd/executor-base/pull/39

### CODE REVIEW

File: test/data/testJob.json
Comment (d2lam) [inline test/data/testJob.json]: should use only 5

File: index.js
Comment (d2lam) [inline index.js]: it will be `beta.screwdriver.cd/buildPeriodically`

**Review by d2lam — CHANGES_REQUESTED** [review]

File: index.js
Comment (d2lam) [inline index.js]: Can you add a comment for while we are doing this. We might not remember later on.

File: index.js
Comment (d2lam) [inline index.js]: Can put `{ pipeline, job}` here

File: index.js
Comment (jithine) [reply index.js]: Why should we have `beta` prefix ?

File: lib/cron.js
Comment (jithine) [inline lib/cron.js]: is this a stub  for now ?

File: index.js
Comment (jithine) [inline index.js]: who invokes `_startPeriodic` can we abstract out all scheduler related code to say `schedule.js` ?

**Review by jithine — CHANGES_REQUESTED** [review]

File: index.js
Comment (d2lam) [inline index.js]: Don't think this will actually work if this is the `config` retrieved back from redis. `admin` is a getter for the `pipeline` object. I don't think you will be able to call `pipeline.admin` here. We might need to just store this `username` directly to redis as a string.

**Review by d2lam — CHANGES_REQUESTED** [review]

File: index.js
Comment (d2lam) [reply index.js]: To be consistent with all the previous annotations. If we want to switch then we should switch for all: https://docs.screwdriver.cd/user-guide/configuration/annotations

File: index.js
Comment (d2lam) [reply index.js]: It invokes from here: https://github.com/screwdriver-cd/models/pull/239/files
I think it makes more sense to be here since it's a method for executor.

File: lib/cron.js
Comment (pranavrc) [reply lib/cron.js]: Yeah I'll update it

File: index.js
Comment (d2lam) [inline index.js]: What about the case where the old job was a periodic job, and was updated to be NOT periodic?
The `models` should still call this function because it needs to stop the periodic. But at this point, we need to check if the config has `buildPeriodically` annotation because the new job might not have it.

**Review by d2lam — CHANGES_REQUESTED** [review]

File: index.js
Comment (minzcmu) [inline index.js]: This can be undefined `config.job.permutations[0].annotations['beta.screwdriver.cd/buildPeriodically`. Need to do some check here

**Review by minzcmu — CHANGES_REQUESTED** [review]

File: lib/transformCron.js
Comment (Filbird) [inline lib/transformCron.js]: Need documentation for this function :)

File: lib/transformCron.js
Comment (Filbird) [inline lib/transformCron.js]: Trivial, but I still believe that `nextExecution` should not directly call `transformCron` and should instead expect a valid cron expression as input (as oppposed to one containing 'H' symbols).

Although this would require adding an additional function call in `_periodicStart`, I think it better separates the responsibilities of both functions.

---

Or we can just use `cron-parser` directly in executor queue.

File: lib/transformCron.js
Comment (Filbird) [inline lib/transformCron.js]: Since we are exporting both `transformCron` and `nextExecution` we can probably generalize the name of the file to something like `cron`.

File: index.js
Comment (jithine) [inline index.js]: why do we need a new `startDelayed` queue, won't `start` work here?

File: lib/transformCron.js
Comment (pranavrc) [reply lib/transformCron.js]: Makes sense, I'll make changes, thanks!

File: index.js
Comment (d2lam) [inline index.js:85]: I think we need to have at least some logging for failure or error, otherwise it will be very hard to debug: https://github.com/screwdriver-cd/queue-worker/blob/master/index.js

Comment (pranavrc) [reply index.js:85]: Added all the logging events for now, will remove based on need

File: index.js
Comment (d2lam) [inline index.js]: Don't think we should have this. It will create a lot of noise in the API log since worker does non-stop polling.

File: lib/cron.js
Comment (Filbird) [inline lib/cron.js]: Can remove `jobId` from parameter and description

File: index.js
Comment (jithine) [inline index.js]: expose a util function and drop `beta`

File: test/index.test.js
Comment (d2lam) [inline test/index.test.js]: For other tests, we call it through the base to make sure the `base` already has this function. Should we do the same here: `executor.startPeriodic`

Comment (pranavrc) [reply test/index.test.js]: Tests will fail until we merge the executor-base PR, but I agree, we should probably merge that first and then this

**Review by minzcmu — COMMENTED** [review]
..moved

**Review by jithine — APPROVED** [review]
👍

File: index.js
Comment (minzcmu) [inline index.js]: Please also check `response.statusCode` before resolve, the behavior is kinda weird. It's possible that err object is empty for non 2XX response.

example:
```
request('https://api.screwdriver.cd/ssss', function (err, response) { console.log(response.statusCode); console.log(error); });
```
statusCode is 404 but error object undefined.

**Review by minzcmu — CHANGES_REQUESTED** [review]

File: index.js
Comment (pranavrc) [reply index.js]: Fixed

File: test/index.test.js
Comment (pranavrc) [reply test/index.test.js]: Fixed

File: index.js
Comment (d2lam) [inline index.js]: This should be a 201?

File: index.js
Comment (d2lam) [inline index.js]: Why do we need this `delete`? Also, good practice is not modifying the param. If you need to modify it then use a different variable.

File: index.js
Comment (minzcmu) [inline index.js]: For /POST event, we return 201
https://github.com/screwdriver-cd/screwdriver/blob/master/plugins/events/create.js#L138

File: index.js
Comment (Zhongtang) [inline index.js]: `triggerBuild` is not invoked by the public function in https://github.com/screwdriver-cd/executor-base. Also this parameter needs to be documented.

File: index.js
Comment (d2lam) [inline index.js]: why we need `delDelayed`?

File: index.js
Comment (minzcmu) [inline index.js:85]: ah... where do we start the worker and scheduler?
https://github.com/screwdriver-cd/queue-worker/blob/master/index.js#L131

**Review by minzcmu — CHANGES_REQUESTED** [review]

File: index.js
Comment (d2lam) [reply index.js]: We can put `triggerBuild` in the config.

File: index.js
Comment (Zhongtang) [inline index.js]: From the implementation of [enqueueAt](https://github.com/taskrabbit/node-resque/blob/1b8509651bf5962b987cce2de819cd1d2c9576b7/lib/queue.js#L65-L66), it seems an exception is thrown if there is an attempt to enqueue the same item with the same timestamp. We may want to catch (and ignore if this is expected) here, so that the ones invoking this function (e.g. model.update()) will be able to proceed?

File: index.js
Comment (pranavrc) [reply index.js:85]: Fixed

File: index.js
Comment (pranavrc) [reply index.js]: I've put it in the config for now so it makes it simpler across the board

File: index.js
Comment (pranavrc) [reply index.js]: Fixed

File: index.js
Comment (pranavrc) [reply index.js]: Changed to use Object.assign

File: index.js
Comment (pranavrc) [reply index.js]: Fixed, included a catch

**Review by d2lam — APPROVED** [review]

**Review by Zhongtang — APPROVED** [review]

**Comment by Zhongtang** [comment]
Not blocking the PR merge, but in the future when we reach higher usage, it might be necessary to synchronize `startDelayed` with the job model's `update()` function.

e.g. a job is scheduled to run (event A), and someone commits a new job config that updates cron (event B) at the same time.
A: hget old cron
A: invokes `startPeriodic`, start build (call `postEvent()`)
B: invokes `startPeriodic` with `isUpdate: true`
B: calls `deldelayed` to remove and redo `enqueueAt` with the new cron information
A: calls `enqueueAt` with the old cron information, and `hset` to update the config back to the old cron

At the end there will be 2 jobs scheduled in the next cycle, and the cron will not be updated.

**Review by minzcmu — APPROVED** [review]

**Pull request merged**
